File: algorithms/sorting/answers/insertion_sort_fp.py
def insert(int_list, key):
    if len(int_list) == 0:
        return [key]
    else:
        if int_list[0] <= key:
            return [key] + int_list
        else:
            return [int_list[0]] + insert(int_list[1:], key)

def iterate(loop_invariant, todo):
    if len(todo) == 0:
        return loop_invariant
    else:
        inserted = insert(loop_invariant, todo[0])
        return iterate(inserted, todo[1:])

def reverse(int_list, container):
    if len(int_list) == 0:
        return container
    else:
        return reverse(int_list[1:], [int_list[0]] + container)

def insertion_sort(int_list):
    if len(int_list) == 0:
        return []
    else:
        return reverse(iterate([int_list[0]], int_list[1:]), [])

def main():
    int_list = [1332, 3,2, 334, 6, 61, 2,4, 245, 567, 220, 2, 2324, 45, 23, 21, 1, 1,3]
    print(int_list)
    print(insertion_sort(int_list))

if __name__ == "__main__":
    main()

File: algorithms/sorting/answers/insertion_sort_fp.ml
let insertion_sort int_list = 
    let rec pour a_list container =
        match a_list with 
        | head :: tail -> pour tail (head :: container)
        | [] -> container in
    let rec insert a_list key =
        match a_list with
        | head :: tail -> 
            if head <= key then key :: a_list
            else head :: (insert tail key)
        | [] -> key :: [] in
    let rec iterate (loop_invariant, todo) =
        match todo with
        | head :: tail ->
            let inserted = insert loop_invariant head in
            iterate (inserted, tail)
        | [] -> loop_invariant in
    match int_list with
    | head :: tail -> pour (iterate (head :: [], tail)) []
    | [] -> []

let rec print_list a_list =
    match a_list with
    | head :: tail ->
        print_int head;
        print_endline "";
        print_list tail
    | [] -> ()

let my_list = [14; 34; 234; 5; 67; 23; 12; 67; 465; 12; 13; 56; 512; 12; 23; 4; 5; 768; 2]

let _ = print_list (insertion_sort my_list)

